### ISSUE TITLE
new word cannot be added to vocabulary by build_vocab

### ISSUE THREAD

**Issue by Atsuyoshi-Funahashi** [piskvorky/gensim#3352]
#### Problem description

Hi, I have a question about the behavior of build_vocab.

I am Japanese and I am using gensim's Word2Vec model by loading the Japanese model [here](http://public.shiroyagi.s3.amazonaws.com/latest-ja-word2vec-gensim-model.zip).

I want to add a new word that is not in the vocabulary, so I created a corpus and tried build_vocab, but about 280 of the 320 or so new words were not registered and I got a key error.

Here is a simple code: build_vocab, train, and then check if the vocab is updated, and I get a keyerorr. I would like to know the cause of this.

```python
import gensim

# model's path
model_path='latest-ja-word2vec-gensim-model/word2vec.gensim.model'

model = gensim.models.Word2Vec.load(model_path)

model.wv["python"]    # error occured　because 'python' is not in the vocab

# adding new word
corpus_list=[["python"]]

# build_vocab and train
model.build_vocab(corpus_list, update=True)
model.train(corpus_list, total_examples=model.corpus_count, epochs=model.epochs)


model.wv["python"]    # error occured

```

#### Versions

```python
gensim version 3.8.1
python version 3.8.9
```

**Comment by gojomo** [comment]
The model likely has a `min_count` setting that automatically discards words that aren't seen at least that many times. That's usually a good policy: these models only work well if many, varied, contrasting uses of a word are available in the training data. Discarding the rarest words with few examples usually *improves* the vectors for the remaining words, & often improves the model's overall performance when used in other downstream tasks. 

You could mutate that parameter down low enough that your words aren't discarded - but of course, that'd mean no longer following a consistent practice with the model's earlier training, and might mean including rare words that, by usual best practices, don't have enough training data to warrant vectors.

You could artificially repeat your data, to fake having enough examples. Because these don't show the same contrasting variety as truly diverse examples, this isn't as good as having more data, but would allow you to evade the normal culling. 

In all cases, beware that such incremental extra-training is at best an advanced/experimental technique, fraught with hidden risks. I've never seen any good writeup explaining effective ways use it in Gensim, with the kind of analysis of considerations/tradeoffs that I'd want to see before using it myself. 

As just a nutshell hint of some of the considerations: imagine you have a corpus `A`, with 900 unique words appearing at least `min_count` times. You train a model, say using the default `epochs=5, min_count=5`. That model fits all usual word2vec assumptions: all words got proportionate training-effort via an equal number of training passes. They're properly sorted in the mdoel by their decreasing frequency. To the extent there were tensions in the A corpus between different word-senses, or a variety of usage domains, all those tensions fought their little-tug-of-war to optimize the final model on equal terms. All is close to published work & standard/best practices. 

Now, you have a smaller corpus `B`, with 200 unique words appearing at least `min_count` times. Let's say 100 of those words are shared with corpus A, and 100 are totally-new words. You want to incrementally train those into your prior model. You now have many new open concerns & questions with no clear answers:

* Your word frequencies from `A` were already truncated at `min_count` - already, words with 4 or fewer occurrences were discarded/ignored. Now, the same happens with `B`. So words that appeared 4 times in `A`, & 4 times in `B`, are ignored - but words that appeared only 5 times in one or the other remain. All the new `B` words get appended at the *end* of the known-words - so the model's words are no longer neatly sorted by true-frequency-in-training data. Some assumptions by users (that they could do a `.most_similar()` on just a N-most-frequent subset of words by the `restrict_vocab` parameter) may now be silently broken. (The efficiency of the negative-sampling mechanism will also be worsened by some higher-frequency words now being at the end.)  
* Should the new batch of training follow the same `alpha` learning-rate decay as the initial training? Then, the 10 `A ⋂ B` words will have been trained with one set of passes going from high-to-low learning-rate, then *another* set of passes also going from high-to-low. Meanwhile the other 90 `A ⊖ B` words each ony get one high-to-low training. Mayb this is OK! But it no longer has either the theoretic grounding in true stochastic-gradient-descent-optimization, nor the similarity to usual practice, that could each create confidence in the results.
* Are 5 more epochs the right choice, such that the 100 `A ⋂ B` words get 10 training passes, the `A`-only words had 5 prior passes but no revisits interleaved (on equal-influence-basis) with `B` words, and the `B`-only words get 5 new passes (but only on equal interleaved basis with other `A ⋂ B` words)? I've never seen any analysis about the right way to choose a value here. Often, a smaller corpus needs more passes... but each pass added for `B` drags all `B` words more into a relative configuration with each other, based only on the new texts, potentially making them arbitrarily-unaligned with any `A` only words (that were only influencing the model arbitrarily-long ago.) How would we detect when `B`-only training has caused too-much drift from `A`-only words? 
* As a generic form of that sort of old-vs-new tensions: what's the risk of '[catastrophic interference](https://en.wikipedia.org/wiki/Catastrophic_interference)' where useful patterns deduced from the original larger training collapse entirely due to the presentation of new training data that no longer refreshes the older learnings? 

The well-grounded choice is to retrain from scratch with data covering all desired vocabulary. Of course, that may take a lot more time, and *definitely* makes the new vectors comparison-ineligible to distinct-prior-run vectors. 

If moving ahead with some other improvised incremental-training/'fine-tuning' approach, you should be aware of the above considerations. You should devise your own criteria for making the open choices, & your own project-appropriate rigorous evaluations to monitor for desired/unwanted effects – because I don't know any good references or best-practices. You'll have the appearance of compatibility with earlier vector-positionings... but such compatibility will be degrading, in hard-to-quantify ways, with each extra deviation from the original model/SGD practices.

**Comment by Sandy4321** [comment]
any updates , is it possible to fine tune unsupervised models ?

**Comment by gojomo** [comment]
People try & do many things that could be described as "fine tuning unsupervised models" in the word2vec/etc space, but:

* I'm not aware of well-proven standard approaches with good tutorials or open-source support
* many online examples just plow through naive approaches without ever checking if they're really helping, or better than other simpler approaches

There's no fresh support in Gensim in recent years. If you have a more-specific goal in mind, where you can describe what you've got & what you're trying to achieve under what constraints, asking with those details on the [project discussion forum ](https://groups.google.com/g/gensim)might get some suggestions.

**Comment by gojomo** [comment]
As the original inquiry here was likely just the `min_count` effect, with no specific bug to fix or feature to add, closing this issue.

**Comment by gojomo** [comment]
As the original inquiry here was likely just the `min_count` effect, with no specific bug to fix or feature to add, closing this issue.

**Comment by gojomo** [comment]
As the original inquiry here was likely just the `min_count` effect, with no specific bug to fix or feature to add, closing this issue.
